userspace/engine/falco_engine.cpp

```diff
@@ -378,24 +384,51 @@ std::unique_ptr<falco_engine::rule_result> falco_engine::process_event(std::size
 		source = find_source(source_idx);
 	}
 
-	if(should_drop_evt() || !source || !source->ruleset->run(ev, source->m_rule, ruleset_id))
+	if(should_drop_evt() || !source)
 	{
-		return std::unique_ptr<struct rule_result>();
+		return nullptr;
 	}
 
-	std::unique_ptr<struct rule_result> res(new rule_result());
-	res->evt = ev;
-	res->rule = source->m_rule.name;
-	res->source = source->m_rule.source;
-	res->format = source->m_rule.output;
-	res->priority_num = source->m_rule.priority;
-	res->tags = source->m_rule.tags;
-	res->exception_fields = source->m_rule.exception_fields;
-	m_rule_stats_manager.on_event(source->m_rule);
+	if (m_rule_matching == falco_common::rule_matching::ALL)
+	{
+		if (!source->ruleset->run(ev, source->m_rules, ruleset_id))
+		{
+			return nullptr;
+		}
+	}
+	else if (m_rule_matching == falco_common::rule_matching::FIRST)
+	{
+		falco_rule rule;
```

Comment:
Let's retain this one too in falco_source. Alternatively, let's make sure that `source->m_rules` always retains 1 element when `m_rule_matching == falco_common::rule_matching::FIRST` and always use that one. Allocating this on the stack at each round is costly, as we found in he past.

Reply:
Makes perfect sense, I am now always retaining 1 element in case of `falco_common::rule_matching::FIRST`